File: client/web/src/features/family/api/selfProfileApiService.ts
import { apiRequest } from '../../../shared/api/apiClient'

export type ProfileRestrictionSeverity = 'STRICT_AVOID' | 'INTOLERANCE'

export interface DietaryRestrictionOption {
  id: number
  code: string
  displayName: string
  category: string
  description?: string | null
}

export interface SelfProfileResponse {
  profileId: number
  profileName: string
  relationship: 'SELF'
  active: boolean
  restrictions: Record<string, ProfileRestrictionSeverity>
}

export const selfProfileEndpoints = {
  catalog: '/api/restrictions',
  create: '/api/profiles/me',
  me: '/api/profiles/me',
} as const

export const selfProfileApiService = {
  getCatalog: () =>
    apiRequest<DietaryRestrictionOption[]>(selfProfileEndpoints.catalog),

  createSelfProfile: (
    profileName: string,
    restrictions: Record<number, ProfileRestrictionSeverity>,
  ) =>
    apiRequest<SelfProfileResponse>(selfProfileEndpoints.create, {
      method: 'POST',
      body: JSON.stringify({ profileName, restrictions }),
      retryAuthentication: false,
    }),

  // Fetches the caller's existing SELF profile so an edit form can be
  // pre-populated. Callers should expect a 404 ApiError when no SELF
  // profile has been created yet.
  getSelfProfile: () =>
    apiRequest<SelfProfileResponse>(selfProfileEndpoints.me),

  updateSelfProfile: (
    profileName: string,
    restrictions: Record<number, ProfileRestrictionSeverity>,
  ) =>
    apiRequest<SelfProfileResponse>(selfProfileEndpoints.me, {
      method: 'PUT',
      body: JSON.stringify({ profileName, restrictions }),
      retryAuthentication: false,
    }),
}

File: client/mobile/app/src/main/java/sg/edu/nus/iss/canmakan/features/dietaryprofile/setup/data/SelfProfileApiService.kt
package sg.edu.nus.iss.canmakan.features.dietaryprofile.setup.data

import com.google.gson.annotations.SerializedName
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.Headers
import retrofit2.http.POST

enum class ProfileRestrictionSeverity {
    STRICT_AVOID,
    INTOLERANCE,
}

data class CreateSelfProfileRequest(
    @SerializedName("profileName") val profileName: String,
    @SerializedName("restrictions")
    val restrictions: Map<Long, ProfileRestrictionSeverity>,
)

data class SelfProfileResponse(
    @SerializedName("profileId") val profileId: Long?,
    @SerializedName("profileName") val profileName: String?,
    @SerializedName("relationship") val relationship: String?,
    @SerializedName("active") val active: Boolean?,
    @SerializedName("restrictions") val restrictions: Map<Long, String>?,
)

interface SelfProfileApiService {
    @Headers("X-CanMakan-No-Retry: true")
    @POST("profiles/me")
    suspend fun createSelfProfile(
        @Body request: CreateSelfProfileRequest,
    ): Response<SelfProfileResponse>
}
